### DISCUSSION TITLE
Minecraft 1.21 Thread

### DISCUSSION THREAD

**Discussion by PokeJofeJr4th** [PokeJofeJr4th/minescript#41] [Announcements]
[Snapshot 23w31a](https://www.minecraft.net/en-us/article/minecraft-snapshot-23w31a) introduces support for multi-version datapacks.  For maximum backwards-compatibility, Minescript will continue to target pack version 15, aka Minecraft 1.20.
## Internal Overlays
Minescript will begin automatically generating overlays whenever it encounters a command set or feature that is different in a future version. Implementation details are still undecided. Minescript may eventually include features that are not possible in earlier datapack versions. If so, it will automatically change the .mcmeta file to reflect the supported datapack versions and produce a warning when compiled.
## Explicit Overlays
Minescript will allow users to explicitly apply version overlays. This is to ensure the user is as able to reach through minescript and write minecraft commands as simply as possible, much like the `@raw` macro. Here's one possible syntax:
```
overlay 16 {
   function my_function {
     @raw "code for datapack version 16"
   }
}

overlay 17 {
  function my_function {
    @raw "code for datapack version 17"
  }
}
```
Ideally, the overlay system shoud allow these to be nested inside other functions, so that a modified version of the function will be overlaid. Here's another potential syntax:
```
@overlay {
  16: @raw "code for version 16"
  17: @raw "code for version 17"
}
```
## Overlay Logistics
Overlay ranges and order should be considered. It's probably best to create some sort of data structure that stores possible versions of different functions. It should be able to collapse versions into a set of overlays.

**Comment by Nico314159** [comment]
The new function macros mean you can create a switch statement that is O(1) instead of O(n). Have each branch be in its own special function and then call `$function switch_case:$(var)` (effectively creating a [jump table](https://en.wikipedia.org/wiki/Branch_table))

**Reply by PokeJofeJr4th** [reply]
I probably won't implement this, since it has to re-parse whenever it's called, that might be slower even though the asymptotic complexity is technically lower. I think the performance hit from that will be worse than the cost of having a longer mcfunction. Edit: I meant to say, this could work as some kind of configuration in case someone really wants to do it.

I do really like the thinking though. Function macros are going to be *insanely* powerful so any creative use is worth considering.

**Reply by Nico314159** [reply]
Nah I did some testing and jump table is way faster than a bunch of if statements, even if you have only ~10 cases

**Reply by Nico314159** [reply]
[Data here](https://docs.google.com/spreadsheets/d/1E9DDz8BFBNM3y0wpujZASnQeWFMTui9YI-Z7wqMm55U/)
